### ISSUE TITLE
Data Storage folder change deletes all data - permission issue

### ISSUE THREAD

**Issue by frederikb96** [osmandapp/OsmAnd#16301]
### Description
If changing the data storage folder from the default one to a manual folder and confirming the move action, all your data gets deleted if you have not the correct permission to access the new folder.

### How to reproduce?
1. Only have the file permission for Media only and not for all files
2. Have your data in the default location like sdcard/Android/data
3. Change the location in the Osmand settings to a manual folder like sdcard/osm
4. Confirm the dialog to move files
5. All files are getting moved and you get a success message and restart app
6. All your data is lost because it got not actually moved because you did not have the permission for the new folder and not got asked for it

### Your Environment
OsmAnd Version: 4.3.5
Android/iOS version: 12
Device model: FP4

**Comment by TheSnoozer** [comment]
Sounds related to https://github.com/osmandapp/OsmAnd/issues/14916, https://github.com/osmandapp/OsmAnd/issues/16043 and https://github.com/osmandapp/OsmAnd/issues/16059.

TLDR: Anroid has become very restrictive with file access permissions for apps.

**Comment by scaidermern** [comment]
> Sounds related to #14916, #16043 and #16059.
> 
> TLDR: Anroid has become very restrictive with file access permissions for apps.

True. However, this is not an excuse for OsmAnd to delete all your data.

OsmAnd should try if the specified storage location is writable before moving files.

**Comment by vshcherb** [comment]
So far we never reproduced the data lost and yes there is a check. Try to change to the previous folder and your data should be back.

**Comment by frederikb96** [comment]
I can confirm that this also happened to me once the first time when changing from data to obb. I was able to change back and my data was still there.
But another time as described in my initial post, I was not able to recover my data. I did immediately revert the storage location and also checked the old location for any files, there were none anymore. I also checked the file system storage and it represented that there was much more available space now, since all data got deleted.

If it is not reproducible, there must be some other factors which let to my situation, which I do not know anymore. So the issue can be closed in that case since at least other users should normally not come into my situation in that case and this was the reason I created the issue.

**Comment by vshcherb** [comment]
Link https://github.com/osmandapp/OsmAnd-Issues/issues/1679

**Comment by Rowin63** [comment]
> I can confirm that this also happened to me once the first time when changing from data to obb. I was able to change back and my data was still there. But another time as described in my initial post, I was not able to recover my data. I did immediately revert the storage location and also checked the old location for any files, there were none anymore. I also checked the file system storage and it represented that there was much more available space now, since all data got deleted.
> 
> If it is not reproducible, there must be some other factors which let to my situation, which I do not know anymore. So the issue can be closed in that case since at least other users should normally not come into my situation in that case and this was the reason I created the issue.

Oh yes, also other users did come to your situation and lose all data. Exactly what happened in my (closed) case #16043.

**Comment by scaidermern** [comment]
> So far we never reproduced the data lost and yes there is a check.

Maybe this check needs to be improved then? For example copy a (dummy) file, then try to read it (and ideally check the size or checksum). Only if this is successful then start to copy the data.

Yes, sounds a little bit exaggerated but it seems like the current check is not enough.

**Comment by vshcherb** [comment]
We need to:
- Create test file and write 4 bytes and close and read and delete it **only if success** start copying process
- Check code that we don't delete file if copying or moving is not true

**Comment by vshcherb** [comment]
Indeed we do investigation how did it happen https://github.com/osmandapp/OsmAnd/pull/15186/commits and how we overlooked in code review / testing. 

I'm sorry it happened and caused the lost of user data, we will proceed with urgent fix.

**Comment by frederikb96** [comment]
Thank you very much :)

**Comment by Zirochkabila** [comment]
OsmAnd~ 4.4.0#14157mqta, released: 2023-02-07

When transferring files to another folder, if there is not enough space in the import folder or not all files have been transferred, an error is displayed informing you that there is not enough space in the folder and that all files could not be transferred.
Only the transferred files are displayed on the map
**But** if you return to the previous type of memory, all files are restored

**Comment by Zirochkabila** [comment]
@frederikb96 can you confirm?

**Comment by frederikb96** [comment]
I  do not know since this is a different szenario
